src/builtin.cpp

```diff
@@ -3028,16 +3028,28 @@ static int builtin_bg(parser_t &parser, io_streams_t &streams, wchar_t **argv) {
             res = send_to_bg(parser, streams, j, _(L"(default)"));
         }
     } else {
-        int i;
-        int pid;
+        std::vector<int> pids;
+
+        // If one argument is not a valid pid (i.e. integer >= 0), fail without backgrounding anything,
+        // but still print errors for all of them.
+        for (int i = 1; argv[i]; i++) {
+            int pid = fish_wcstoi(argv[i]);
+            if (errno || pid < 0) {
+                streams.err.append_format(_(L"%ls: '%ls' is not a valid job specifier\n"), argv[0], argv[i]);
+                res = STATUS_BUILTIN_ERROR;
+            }
+            pids.push_back(pid);
+        }
+        if (res == STATUS_BUILTIN_ERROR) {
+            return res;
+        }
 
-        for (i = 1; argv[i]; i++) {
-            pid = fish_wcstoi(argv[i]);
-            if (errno || pid < 0 || !job_get_from_pid(pid)) {
-                streams.err.append_format(_(L"%ls: '%ls' is not a job\n"), argv[0], argv[i]);
-                return STATUS_BUILTIN_ERROR;
+        // Background all existing jobs that match the pids.
+        // Non-existent jobs aren't an error.
+        for (auto p : pids) {
+            if (job_t* j = job_get_from_pid(p)) {
```

Comment:
This seems slightly schizophrenic. We don't display an error msg or return a non-zero status if the the pid doesn't match a known job. But we do if we can't send it to the background? If you look at `send_to_bg()` you'll see the original author felt a non-existent job should be an error. It seems to me this should unconditionally pass the result from `job_get_from_pid()` to `send_to_bg()` and let it report an error if the job lookup failed.